File: backend/src/models/invoice.rs
use chrono::{DateTime, Utc};
use serde::{Serialize,Deserialize};
use rust_decimal::Decimal;


#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct Invoice {
    pub id: String,
    pub flag: InvoiceFlags,
    pub invoice_number: String,
    pub username: String,
    pub company_tin: String,
    pub client_name: String,
    pub client_tin: String,
    pub invoice_date: DateTime<Utc>,
    pub invoice_time: DateTime<Utc>,
    pub due_date: DateTime<Utc>,
    #[sqlx(default)]
    pub total_vat: Decimal, 
    #[sqlx(default)]
    pub total_amount: Decimal,
    pub status: InvoiceStatus,
    pub items: Vec<InvoiceItem>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct InvoiceItem {
    pub id: String,
    pub invoice_id: String, // foreign key to Invoice
    pub item_code: String,    // foreign key to Item
    pub quantity: i32,
    pub unit_price: Decimal,
    pub total_vat: Decimal,
    pub total_levy: Decimal,
    pub total_discount: Decimal,
    pub line_total: Decimal, // (quantity * unit_price) - discount + tax
}

#[derive(Debug, Serialize, Deserialize, sqlx::Type)]
pub enum InvoiceFlags {
    Invoice,
    Purchase,
    PartialRefund,
    FullRefund,
    PurchaseReturn,
    CreditNote,
}

#[derive(Debug, Serialize, Deserialize, sqlx::Type)]
pub enum InvoiceStatus {
    Draft,
    Sent,
    Paid,
    Overdue,
    Canceled,
}

impl std::fmt::Display for InvoiceFlags {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            InvoiceFlags::Invoice => write!(f, "Invoice"),
            InvoiceFlags::Purchase => write!(f, "Purchase"),
            InvoiceFlags::PartialRefund => write!(f, "Partial Purchase"),
            InvoiceFlags::FullRefund => write!(f, "Full Purchase"), 
            InvoiceFlags::PurchaseReturn => write!(f, "Purchase Return"),
            InvoiceFlags::CreditNote => write!(f, "Credit Note"),
        }
    }
}

File: server/internal/models/invoice.go
package models

import "time"

type Invoice struct {
	ID                  string `gorm:"primaryKey;type:char(36)"`
	InvoiceNumber       string `json:"invoiceNumber"`
	Username            string `json:"username"`
	BusinessPartnerName string
	BusinessPartnerTIN  string          `json:"businessPartnerTIN"`
	BusinessPartner     Client          `gorm:"foreignKey:BusinessPartnerTIN;references:ClientTIN" json:"-"`
	InvoiceType         InvoiceType     `json:"flag"`
	CalculationType     CalculationType `json:"calculationType"`
	InvoiceDate         time.Time       `json:"invoiceDate"`
	InvoiceTime         time.Time
	DueDate             time.Time
	TotalVAT            float64 `json:"totalVAT"`
	TotalAmount         float64 `json:"totalAmount"`
	CompanyTIN          string
	Company             Company       `gorm:"foreignKey:CompanyTIN;references:TIN" json:"-"`
	InvoiceItems        []InvoiceItem `gorm:"foreignKey:InvoiceID;references:ID"`
}

type InvoiceItem struct {
	ID          string `gorm:"primaryKey;type:char(36)"`
	InvoiceID   string `gorm:"type:char(36)"`
	ItemID      string `gorm:"type:char(36)"`
	Description string
	Quantity    float64
	UnitPrice   float64
	Total       float64
}

type InvoiceType string

const (
	SalesInvoice InvoiceType = "Invoice"
	Purchase     InvoiceType = "Purchase"
	Refund       InvoiceType = "Refund"
	CreditNote   InvoiceType = "Credit Note"
	DebitNote    InvoiceType = "Debit Note"
	// PurchaseReturn
)

type CalculationType string

const (
	Inclusive CalculationType = "INCLUSIVE"
	Exclusive CalculationType = "EXCLUSIVE"
)
